pandas/tests/tools/test_concat.py

```diff
@@ -1899,3 +1899,64 @@ def test_concat_multiindex_dfs_with_deepcopy(self):
         tm.assert_frame_equal(result_copy, expected)
         result_no_copy = pd.concat(example_dict, names=['testname'])
         tm.assert_frame_equal(result_no_copy, expected)
+
+    def test_concat_no_unnecessary_upcats(self):
+        # GH 13247
+        for pdt in [pd.Series, pd.DataFrame, pd.Panel, pd.Panel4D]:
+            dims = pdt().ndim
+            for dt in np.sctypes['float']:
+                dfs = [pdt(np.array([1], dtype=dt, ndmin=dims)),
+                       pdt(np.array([np.nan], dtype=dt, ndmin=dims)),
+                       pdt(np.array([5], dtype=dt, ndmin=dims))]
+                x = pd.concat(dfs)
+                self.assertTrue(x.values.dtype == dt)
+
+            for dt in (np.sctypes['int'] + np.sctypes['uint']):
+                dfs = [pdt(np.array([1], dtype=dt, ndmin=dims)),
+                       pdt(np.array([5], dtype=dt, ndmin=dims))]
+                x = pd.concat(dfs)
+                self.assertTrue(x.values.dtype == dt)
+
+            objs = []
+            objs.append(pdt(np.array([1], dtype=np.float32, ndmin=dims)))
+            objs.append(pdt(np.array([1], dtype=np.float16, ndmin=dims)))
+            self.assertTrue(pd.concat(objs).values.dtype == np.float32)
+
+            objs = []
+            objs.append(pdt(np.array([1], dtype=np.int32, ndmin=dims)))
+            objs.append(pdt(np.array([1], dtype=np.int64, ndmin=dims)))
+            self.assertTrue(pd.concat(objs).values.dtype == np.int64)
+
+            # not sure what is the best answer here
+            objs = []
+            objs.append(pdt(np.array([1], dtype=np.int32, ndmin=dims)))
+            objs.append(pdt(np.array([1], dtype=np.float16, ndmin=dims)))
+            self.assertTrue(pd.concat(objs).values.dtype == np.float64)
+
+    @parametrize('dtype', np.sctypes('float'))
+    @parametrize('klass', [Series, DataFrame, Panel])
+    def test_concat_no_unnecessary_upcats_pytest(self, dtype, klass):
+        # GH 13247
+        for pdt in klass:
```

Comment:
I could be misreading this, but I don't think you need the for loop over `klass` anymore. pytest will handle that for you, since you're doing the parametrize.
I think this will also fail, since this class eventually inherits from `TestCase`, which doesn't play well with parametrized fixtures. I'd try it like this:

```python
import numpy as np
from pandas import Series, DataFrame, Panel
import pandas as pd
import pytest


class TestConcatUpcast(object):

    @pytest.mark.parametrize('pdt', [Series, DataFrame, Panel])
    def test_concat_no_unnecessary_upcats_pytest(self, pdt):
        # GH 13247
        dims = pdt().ndim
        for dt in np.sctypes['float']:
            dfs = [pdt(np.array([1], dtype=dt, ndmin=dims)),
                   pdt(np.array([np.nan], dtype=dt, ndmin=dims)),
                   pdt(np.array([5], dtype=dt, ndmin=dims))]
            x = pd.concat(dfs)
            assert x.values.dtype == dt

            objs = []
            objs.append(pdt(np.array([1], dtype=np.float32, ndmin=dims)))
            objs.append(pdt(np.array([1], dtype=np.float16, ndmin=dims)))
            assert pd.concat(objs).values.dtype == np.float32

            objs = []
            objs.append(pdt(np.array([1], dtype=np.int32, ndmin=dims)))
            objs.append(pdt(np.array([1], dtype=np.int64, ndmin=dims)))
            assert pd.concat(objs).values.dtype == np.int64

            objs = []
            objs.append(pdt(np.array([1], dtype=np.int32, ndmin=dims)))
            objs.append(pdt(np.array([1], dtype=np.float16, ndmin=dims)))
            assert pd.concat(objs).values.dtype == np.float64
```

This doesn't parametrize all that well over the `dtype` argument, so probably best to just leavve it in there.

If you want to leave comments on https://github.com/pandas-dev/pandas/issues/15608, would appreciate your insight.